### ISSUE TITLE
Display hooks state in devtools

### ISSUE THREAD

**Issue by marvinhagemeister** [preactjs/preact-devtools#52]
It's not as straightforward as I initially hoped, because `fiber` specific code is leaking into the `Agent`:

https://github.com/facebook/react-devtools/blob/39072b1bf124c0a934c42ee970db3832afc27ac8/plugins/HooksInspector/backend.js#L63-L67

**Comment by MaxmaxmaximusGitHub** [comment]
you kist just must create TWO versions of library =)
dev version (widh captures and warnings) and production version (with optimissations)

no?)

**Comment by marvinhagemeister** [comment]
Note that this is not related to hooks, which this issue here is about.

There is an even more elegant way to add debugging warnings that we chose for Preact: The core package has an additional `preact/debug` import which injects debugging messages. This approach has several advantages over conditional dev-only code in the same package:

- Testing setup is much easier, no need to run same tests twice
- Best for smallest size due to tree-shaking
- Can be easily added in no-build-tools environments. Just add a script tag and boom, full debugging capabilities with Preact

**Comment by MaxmaxmaximusGitHub** [comment]
ok, I just understood your rhetoric above as "we can’t add features, since we are focused on optimization," I answered this way)

**Comment by marvinhagemeister** [comment]
> ok, I just understood your rhetoric above as "we can’t add features, since we are focused on optimization," I answered this way)

No, that's not what I'm saying. Rather that we _already_ have built this in `preact` itself :+1: 

> we can detect dom elements as components

yup, that's done by checking for `vnode.type`. If it's of type:
  1.1 `null` -> Text node
  1.2 `string` -> DOM node
  1.3 `function` -> function component
  1.4 `function` with prototype -> class component

> I am so inspired by your project that I can even help you, just tell me which files you need to edit and I'll do everything by morning ^ __ ^

Sounds great! A good way to start is probably to look at #83 .

> preact is what made me brain, at first I thought that 3 kilobytes is a joke or a typo)) but for the last 2 days I can’t fully realize that you, alone, have won a billionth corporation

I'm happy to hear that Preact makes you excited about web development. We do take our 3kb stance pretty seriously and try to keep code as simple as possible given the constraints. Note that this is requires team effort. The devtools extension and inspection features are mostly written by my at the time of this writing, but Preact itself enjoys the contributions from various people. It wouldn't be where it is today without them and our awesome userbase :100: 

## Hooks inspection vs state/props inspection

The way hooks work internally is in stark contrast to `props` or `state`. For the latter we can just take the `object`, serialize it and send it to the devtools. Hooks on the other hand need a lot more hand holding.

The data you see in the sidebar of the devtools is only retrieved on demand. Back when we attached Preact to the React devtools V3 this wasn't the case. Every information about every node was collected upfront. This led to a very noticeable slow down, even on fast machines.

To combat that both the React devools and Preact devtools only collect the information that's displayed in the sidebar when needed. When we want to inspect hooks we have access to the component and Preact's internal structures where the necessary state for hooks is stored. So we can get the value of the `useState` calls below pretty easily.

```js
function MyComponent() {
  const [v1, set1] = useState(0);
  const [v2, set2] = useState(0);
  ...
}
```

The inspection of the above component should show up in the devtools as something like this:

- useState: 0
- useState: 0

Both elements have the same name, so we need a parser that can deals with that. `Props` and `State` are a standard JavaScript object, where something like that can't happen. Whenever there are duplicate property names, only the last one will be picked.

```js
const obj = {
  foo: 1,
  foo: 2,
}

console.log(obj);
// Logs: { foo: 2 }
```

So we need a different underlying data structure represent hooks state. Most of the preparations of that are in place and the last pieces will come with #143 .

The next tricky bit is detecting custom hooks. This is something unique to hooks and not needed to display a component's state or props in the devtools. A custom hook is essentially composed of nested function calls:

```js
function useCounter() {
  return useState(0);
}

function useFancyCounter() {
  return useCounter()
}

function MyComponent() {
  const [v, set] = useFancyCounter();
  ...
}
```

When a "native" hook like `useState` is called we want to know which custom hooks it was part of. Remember that we collect this information lazily. The internal data structures only show that a `useState` hook with the value `0` was used for `MyComponent`. We don't know anything about `useCounter` or `useFancyCounter` at this point. We roughly want to display something like this in the devtools:

- useFancyCounter
  - useCounter
    - useState: 0

But how do we know what function our native hook was called in? Turns out that there isn't a nice way to do that in most programming languages, including JavaScript. We can't leverage a custom compiler or babel plugin that would insert custom statements or wrap these functions as Preact is often used without any build tooling at all.

Luckily we are only interested in the name of the caller functions and don't need anything else. We can get that by throwing an error and looking at the stack trace. It's the only way to do that in JavaScript.

To get that stack trace we need to re-execute the component so that our hooks functions will be called again. But it would very weird for users when inspecting a component, so we need to make sure that no application state is modified.

The whole virtual-dom idea relies on render functions to be pure and hooks push the effectful code into `useEffect` and `useLayoutEffect` respectively. By adding a custom code path in Preact we can bypass all effect invocations whenever we re-render the inspected components to collect the hooks data. Once re-rendered we can create an `Error` object inside `useState` and pull off the stack trace from it.

A stack trace in Chrome looks something like this:

```txt
Error: fail
    at useState (<anonymous>:1:33)
    at useCounter (<anonymous>:1:19)
    at useFancyCounter (<anonymous>):1:13
    at MyComponent
```

_Just an example, line numbers are not real here and we often get the file name too instead of `anonymous`_

Since there are competing browsers used in the wild we need to take different formats into account. The have slightly different formatting, but the parsing is generally quite easy to do.

With that out of the way we have all the information we need, just in the wrong format. If we do the inspection on the following component:

```js
function useFoo() {
  const a = useState(0);
  const b = useState(0);
  return { a, b }
}

function MyComponent() {
  const foo = useFoo();
  const foo2 = useFoo();
  ...
}
```

we will end up with something like this once we parsed the stack trace:

```js
[
  // First useState stack trace
  [
    { name: "useState", lineNumber: ..., column: ..., fileName: ... },
    { name: "useFoo", lineNumber: ..., column: ..., fileName: ... },
  ],
  // Second useState stack trace
  [
    { name: "useState", lineNumber: ..., column: ..., fileName: ... },
    { name: "useFoo", lineNumber: ..., column: ..., fileName: ... },
  ],
  // Third useState stack trace
  [
    { name: "useState", lineNumber: ..., column: ..., fileName: ... },
    { name: "useFoo", lineNumber: ..., column: ..., fileName: ... },
  ],
  // Fourth useState stack trace
  [
    { name: "useState", lineNumber: ..., column: ..., fileName: ... },
    { name: "useFoo", lineNumber: ..., column: ..., fileName: ... },
  ],
]
```

We need to do some heuristics on the parsed stack trace to detect the correct hierarchy of the hooks. We can do that by matching the file name, line and column numbers.

Once we have done this, we can finally send this over to the devtools. All this is a special kind of madness, but without a compiler inserting custom debugging code, there is not much else one can do. The React devtools employ the same strategy.

To sum it up the difference between displaying props, state and hooks is this:

props/state:

  1. Serialize standard JavaScript object
  2. Send to devtools

hooks:

  1. Re-render component. Don't forget to disable any effects
  2. Collect stack traces from each native hook
  3. Parse said stack traces
  4. Turn parsed stack traces into a tree hierarchy
  5. Serialize tree
  6. Send to devtools

So yeah, inspecting hooks is a bit more complex. It's doable and I'm close to completing it, but I need a few more days :tada:
